src/GOWIN.py

```diff
@@ -206,42 +386,57 @@ def SYN_AND_REPORT_TIMING_NEW(
         parser_state, inst_name
     )
 
-    # Generate tcl file TODO use real generated vhdl files
-    tcl_txt = ""
-    tcl_txt += f'''
-set_device -name {parser_state.part}
-'''
-    # All the generated vhdl files
-    for vhdl_file_text in vhdl_files_texts.split(" "):  # hacky
-        vhdl_file_text = vhdl_file_text.strip()
-        if vhdl_file_text != "":
-            tcl_txt += (
-                """add_file -type vhdl """
-                + vhdl_file_text
-                + "\n")
-    tcl_txt += f'''
+    # Which VHDL files?
+    vhdl_files_texts, top_entity_name = SYN.GET_VHDL_FILES_TCL_TEXT_AND_TOP(
+        multimain_timing_params, parser_state, inst_name
+    )
+
+    # Generate TCL file
+    tcl_file = top_entity_name + ".tcl"
+    tcl_path = output_directory + "/" + tcl_file
+
+    # parse part name to direct the synthesis 
+    part_name_parts = parser_state.part.split(":")
+    if len(part_name_parts) > 1:
+        part_name, device_version = part_name_parts
+    else:
+        part_name = part_name_parts[0]
+        device_version = None
+    tool_part_name, device_versions = PART_TO_DEVICE_VERSIONS(part_name)
+
+    with open(tcl_path, "w") as f:
+        if device_version:
+            f.write(f"set_device --device_version {device_version} {tool_part_name}\n")
+        elif len(device_versions) > 1:
+            f.write(f"set_device --device_version {device_versions[-1]} {tool_part_name}\n")
+        else:
+            f.write(f"set_device {tool_part_name}\n")
+        for vhdl_file_text in vhdl_files_texts.split(" "):
+            vhdl_file_text = vhdl_file_text.strip()
+            if vhdl_file_text != "":
+                f.write(f"add_file -type vhdl {vhdl_file_text}\n")
+        f.write(
+            f'''
 add_file -type sdc {constraints_filepath}
-set_option -output_base_name pipelinec
+set_option -output_base_name {top_entity_name}
 set_option -top_module {top_entity_name}
 set_option -vhdl_std vhd2008
 set_option -gen_text_timing_rpt 1
-run all
-'''
-    tcl_file = top_entity_name + ".tcl"
-    tcl_path = output_directory + "/" + tcl_file
-    f = open(tcl_path, "w")
-    f.write(tcl_txt)
-    f.close()
-
-    # Run the build script
-    print("Running Gowin:", tcl_path, flush=True)
-    syn_imp_bash_cmd = (
-        GOWIN_PATH + "/" + TOOL_EXE + " " + tcl_path
-    )
+run {"all" if DO_PNR is "all" else "syn"} 
```

Comment:
Ooo niiice using the DO_PNR config here like vivado too - looks like the html report is parsed the same both ways too - right on :muscle: